### PR TITLE
[WR122] Update Realm Group

### PR DESCRIPTION
**Pull request by house9**

### CODE REVIEW

**Comment by house9** [comment]
Left out the changes to `dist` for now, npm package is going to include a few more tickets

File: src/api/realmGroups.ts [hunk 1]
```diff
@@ -4,7 +4,22 @@ import { APIContext } from './context'
...
+// transformAttributesForApi converts attributes to expected api format
+// { key1: 'value1' } becomes { key1: ['value1']}
+function transformAttributesForApi(payload: any) {
+  const attributes: any = {}
+  for (const [key, value] of Object.entries(payload.attributes || {})) {
+    attributes[key] = [value]
+  }
+  payload.attributes = attributes
+
+  return payload
```
Comment (house9) [inline src/api/realmGroups.ts:22]: I am basically by-passing the typing by using `any`. The `attributes` day is dynamic in nature so hopefully this is ok.

I had

```
type ApiAttribute = { [prop: string]: [any] }

function transformAttributesForApi(payload: any) {
  const attributes: ApiAttribute = {}
  ...
```

but it seemed like overkill to me?

**Review by efabens — APPROVED** [review]

File: src/api/realmGroups.ts (hunk 1 above)
Comment (pirtleshell) [reply src/api/realmGroups.ts:22]: i'm not familiar with group attrs. are they always a map of strings to an array of strings? or can the values be something else?

i'd probably use `Record<string, string>` or `Record<string, string[]>` despite the overkill

though there are enough `any`s floating around that it doesnt' make too much a difference 😆  i'd more prefer the explicit typing for documenting to future me what `attributes` are.

File: src/api/realmGroups.ts [hunk 2]
```diff
@@ -4,7 +4,22 @@ import { APIContext } from './context'
 
 type CreateRealmGroupData = {
   realmName: string
-  group: { name: string }
+  group: {
+    name: string
+    attributes: object
```
Comment (pirtleshell) [inline src/api/realmGroups.ts:9]: probably make optional with `attributes?:`

note on type in other comment! 😁

File: src/types/group.ts [hunk 3]
```diff
@@ -5,11 +5,13 @@ class Group {
   id: string
   name: string
   path: string
+  attributes: { [prop: string]: any }
```
Comment (pirtleshell) [inline src/types/group.ts:8]: i would declare and export a `RealmGroupAttributes` type in this file (even if just as an alias for `any`) to use both here and in the api functions to reduce tacit duplication of those type defs.

**Review by pirtleshell — APPROVED** [review]
approved pending `attributes` typing 🎉

File: src/api/realmGroups.ts (hunk 1 above)
Comment (house9) [reply src/api/realmGroups.ts:22]: On the server it is `key:string, value:string[]` but in the client we want the users to just use `key: string, value:string`
Keycloak uses an array of string values for some reason, even though their UI just supports a single value.

I'll give `Record<string, string>` a go, thanks.

**Review by galxy25 — APPROVED** [review]
approve pending resolution of @pirtleshell excellent comments

**Pull request merged**